cmd/mqtt-log-stdout/main.go

```diff
@@ -47,41 +52,73 @@ func start(cfg config.Client) error {
 	metricsServer := newMetricsServer(cfg, statusClient)
 	mqttClient := newMqttClient(cfg, statusClient, messageClient)
 
-	go metricsServer.Start()
+	start(ctx, g, metricsServer)
+	start(ctx, g, mqttClient)
 
-	var result error
-	err := mqttClient.Start()
-	if err != nil {
-		statusClient.Print("Received error starting mqtt client", err)
-		result = multierror.Append(result, err)
-	}
+	stoppedBy := waitForStop(stopChan, ctx)
+	statusClient.Print(fmt.Sprintf("Application stopping, initiated by: %s", stoppedBy), nil)
 
-	stoppedBy := func() string {
-		select {
-		case sig := <-stopChan:
-			return fmt.Sprintf("os.Signal (%s)", sig)
-		case <-mqttClient.Done():
-			return "mqtt client"
-		case <-metricsServer.Done():
-			return "metrics server"
-		}
-	}()
+	cancel()
 
-	statusClient.Print(fmt.Sprintf("Application stopping, initiated by: %s", stoppedBy), nil)
+	timeoutCtx, timeoutCancel := newTimeoutContext()
+	defer timeoutCancel()
+
+	stop(timeoutCtx, g, mqttClient)
+	stop(timeoutCtx, g, metricsServer)
 
-	err = mqttClient.Stop()
+	return waitForErrGroup(g)
```

Comment:
This one is over my head. I understand your logic and need to read up on it some more. Sounds logical to have two different groups for this, but at the same time we are using different contexts. Not sure what is the "correct" way here...

Reply:
After looking at this deeper, and building a test to verify my logic, I've come to the conclusion that I want the same errgroup since I want the `Wait()` to return an error if any of the `Start()` or `Stop()` returns an error.